aeneas/src/ir/VariantSolver.v3

```diff
@@ -83,6 +101,197 @@ def ON_STACK = -1;
 def EMPTY_FIELD = VariantField.new(null, []);
 def NO_FIELDS = Array<VariantField>.new(0);
 
+class VariantSolver(vnorm: VariantNormalizer, usePacking: bool, useTagScalar: bool) {
+	var normFields: Array<Array<Type>>;
+
+	var fields: Vector<CaseField>; // an ordering of fields to solve for during packing
+
+	// final solutions
+	var scalarAsns: HashMap<CaseField, int>; // field location -> scalar idx
+	var intervalAsns: HashMap<CaseField, Interval>; // field location -> interval within scalar
+	var state: Array<VariantPattern>;
+
+	var explicitTagLength: byte;
+	var explicitTag: (int, Interval) = (-1, Interval(0, 0));
+
+	var curRep: Vector<Scalar.set>;
+	var curUsed: Vector<byte>;
+	
+	// Perform recursive backtracking on the potential representation.
+	// With each scalar in each case, we have several choices to make:
+	// 1. Pack this scalar with an existing scalar
+	// 2. Append this scalar as a new scalar.
+	// We can only determine distinguishability after all the scalars have been assigned. Heuristics will help speed this part up.
+	private def tryRepresentationForField(curCase: int, curField: int) -> bool {
```

Comment:
It's hard to follow how this works because this an internal part of the recursion; the overall recursive algorithm should come first and be documented; then each individual part should be documented, along with the arguments.